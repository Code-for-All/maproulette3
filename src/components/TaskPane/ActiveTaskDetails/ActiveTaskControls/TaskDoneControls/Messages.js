import { defineMessages } from 'react-intl'

/**
 * Internationalized messages for use with TaskDoneControls.
 */
export default defineMessages({
  completedPrompt: {
    id: 'ActiveTask.completed.prompt',
    defaultMessage: "Completed! Marked as",
  },

  nextTask: {
    id: 'ActiveTask.controls.nextTask.label',
    defaultMessage: "Next Task",
  },
})
