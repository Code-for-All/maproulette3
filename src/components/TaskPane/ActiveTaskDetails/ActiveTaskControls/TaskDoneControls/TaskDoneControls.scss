@import '../../../../../variables.scss';

.task-done-controls {
  display: flex;
  flex-direction: column;
  align-items: center;

  &__next {
    margin-top: 15px;
    margin-bottom: 15px;
    font-size: 18px;

    svg {
      height: 18px;
      width: auto;
      fill: $primary;
      vertical-align: middle;
      margin-left: 0.25em;
    }

    &:hover, &:enabled:focus {
      color: $green;

      svg {
        fill: $green;
      }
    }
  }


  &.is-minimized {
    .task-done-controls__next {
      margin-top: 0;
    }
  }
}
