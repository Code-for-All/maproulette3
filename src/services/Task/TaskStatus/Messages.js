import { defineMessages } from 'react-intl'

/**
 * Internationalized messages for use with TaskStatus.
 */
export default defineMessages({
  created: {
    id: "Task.status.created",
    defaultMessage: "Created"
  },
  fixed: {
    id: "Task.status.fixed",
    defaultMessage: "Fixed"
  },
  falsePositive: {
    id: "Task.status.falsePositive",
    defaultMessage: "Not an Issue"
  },
  skipped: {
    id: "Task.status.skipped",
    defaultMessage: "Skipped"
  },
  deleted: {
    id: "Task.status.deleted",
    defaultMessage: "Deleted"
  },
  alreadyFixed: {
    id: "Task.status.alreadyFixed",
    defaultMessage: "Already Fixed"
  },
  tooHard: {
    id: "Task.status.tooHard",
    defaultMessage: "Too Hard"
  },
})
