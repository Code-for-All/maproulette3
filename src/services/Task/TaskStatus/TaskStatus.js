import { map as _map,
         invert as _invert,
         fromPairs as _fromPairs } from 'lodash'
import messages from './Messages'

// These statuses are defined on the server
export const TASK_STATUS_CREATED = 0
export const TASK_STATUS_FIXED = 1
export const TASK_STATUS_FALSE_POSITIVE = 2
export const TASK_STATUS_SKIPPED = 3
export const TASK_STATUS_DELETED = 4
export const TASK_STATUS_ALREADY_FIXED = 5
export const TASK_STATUS_TOO_HARD = 6

export const TaskStatus = Object.freeze({
  created: TASK_STATUS_CREATED,
  fixed: TASK_STATUS_FIXED,
  falsePositive: TASK_STATUS_FALSE_POSITIVE,
  skipped: TASK_STATUS_SKIPPED,
  deleted: TASK_STATUS_DELETED,
  alreadyFixed: TASK_STATUS_ALREADY_FIXED,
  tooHard: TASK_STATUS_TOO_HARD,
})

export const keysByStatus = Object.freeze(_invert(TaskStatus))

/**
 * Returns true if the given status represents completion of
 * the task.
 */
export const isCompleted = function(status) {
  return status === TaskStatus.fixed ||
         status === TaskStatus.falsePositive ||
         status === TaskStatus.alreadyFixed ||
         status === TaskStatus.deleted
}

/**
 * Returns an object mapping difficulty values to raw internationalized
 * messages suitable for use with FormattedMessage or formatMessage.
 */
export const messagesByStatus = _fromPairs(
  _map(messages, (message, key) => [TaskStatus[key], message])
)

/** Returns object containing localized labels  */
export const statusLabels = intl => _fromPairs(
  _map(messages, (message, key) => [key, intl.formatMessage(message)])
)
