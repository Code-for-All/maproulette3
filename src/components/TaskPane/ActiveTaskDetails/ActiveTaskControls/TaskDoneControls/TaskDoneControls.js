import React, { Component } from 'react'
import PropTypes from 'prop-types'
import classNames from 'classnames'
import { FormattedMessage } from 'react-intl'
import { messagesByStatus,
         isCompleted } from '../../../../../services/Task/TaskStatus/TaskStatus'
import SvgSymbol from '../../../../SvgSymbol/SvgSymbol'
import messages from './Messages'
import './TaskDoneControls.css'

/**
 * TaskDoneControls presents controls for interacting with a task that has
 * already been marked with some sort of completed status. It displays the
 * existing status and at least a button for moving on to another next task.
 *
 * @author [Neil Rotstan](https://github.com/nrotstan)
 */
export default class TaskDoneControls extends Component {
  render() {
    if (!this.props.task || !isCompleted(this.props.task.status)) {
      return null
    }

    const statusMessage =
      this.props.intl.formatMessage(messages.completedPrompt) +
      ' ' +
      this.props.intl.formatMessage(messagesByStatus[this.props.task.status])

    return (
      <div className={classNames("task-done-controls",
                                 {"is-minimized": this.props.isMinimized})}>
        {!this.props.isMinimized &&
          <div className="task-done-controls--task-status-message">
            {statusMessage}
          </div>
        }

        <div className="has-centered-children">
          <a className="is-text task-done-controls__next"
             title={this.props.intl.formatMessage(messages.nextTask)}
             onClick={() => this.props.nextTask(this.props.task.parent.id, this.props.task.id)}>
            {!this.props.isMinimized && <FormattedMessage {...messages.nextTask} />}
            <SvgSymbol viewBox='0 0 20 20' sym="forward-icon" />
          </a>
        </div>
      </div>
    )
  }
}

TaskDoneControls.propTypes = {
  /** The current task */
  task: PropTypes.object,
  /** Invoked if the user wishes to move to a new task */
  nextTask: PropTypes.func.isRequired,
}
