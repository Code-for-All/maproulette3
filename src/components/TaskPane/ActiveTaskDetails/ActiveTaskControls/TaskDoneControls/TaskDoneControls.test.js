import React from 'react'
import { omit as _omit, cloneDeep as _cloneDeep } from 'lodash'
import TaskDoneControls from './TaskDoneControls'
import { TaskStatus } from '../../../../../services/Task/TaskStatus/TaskStatus'

const propsFixture = {
  task: {
    id: 123,
    status: TaskStatus.fixed,
    parent: {
      id: 321,
    }
  },
  user: {
    id: 357,
    settings: {defaultEditor: 1},
    isLoggedIn: true,
  },
}

let basicProps = null

beforeEach(() => {
  basicProps = _cloneDeep(propsFixture)

  basicProps.nextTask = jest.fn()
  basicProps.intl = {formatMessage: jest.fn()}
})

test("it renders the completed status of the task", () => {
  const wrapper = shallow(
    <TaskDoneControls {...basicProps} />
  )

  expect(wrapper.find('.task-done-controls--task-status-message').exists()).toBe(true)

  expect(wrapper).toMatchSnapshot()
})

test("clicking the next button signals that a new task should be loaded", () => {
  const wrapper = shallow(
    <TaskDoneControls {...basicProps} />
  )

  wrapper.find('.task-done-controls__next').simulate('click')

  expect(basicProps.nextTask.mock.calls.length).toBe(1)
  expect(basicProps.nextTask.mock.calls[0][0]).toBe(basicProps.task.parent.id)
  expect(basicProps.nextTask.mock.calls[0][1]).toBe(basicProps.task.id)
})
