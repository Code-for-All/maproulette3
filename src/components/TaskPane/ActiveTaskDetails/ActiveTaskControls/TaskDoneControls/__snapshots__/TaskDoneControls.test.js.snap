// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it renders the completed status of the task 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <TaskDoneControls
    intl={
        Object {
            "formatMessage": [Function],
          }
    }
    nextTask={[Function]}
    task={
        Object {
            "id": 123,
            "parent": Object {
              "id": 321,
            },
            "status": 1,
          }
    }
    user={
        Object {
            "id": 357,
            "isLoggedIn": true,
            "settings": Object {
              "defaultEditor": 1,
            },
          }
    }
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <div
          className="task-done-controls--task-status-message"
>
          undefined undefined
</div>,
        <div
          className="has-centered-children"
>
          <a
                    className="is-text task-done-controls__next"
                    onClick={[Function]}
                    title={undefined}
          >
                    <FormattedMessage
                              defaultMessage="Next Task"
                              id="ActiveTask.controls.nextTask.label"
                              values={Object {}}
                    />
                    <SvgSymbol
                              sym="forward-icon"
                              viewBox="0 0 20 20"
                    />
          </a>
</div>,
      ],
      "className": "task-done-controls",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "undefined undefined",
          "className": "task-done-controls--task-status-message",
        },
        "ref": null,
        "rendered": "undefined undefined",
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <a
            className="is-text task-done-controls__next"
            onClick={[Function]}
            title={undefined}
>
            <FormattedMessage
                        defaultMessage="Next Task"
                        id="ActiveTask.controls.nextTask.label"
                        values={Object {}}
            />
            <SvgSymbol
                        sym="forward-icon"
                        viewBox="0 0 20 20"
            />
</a>,
          "className": "has-centered-children",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <FormattedMessage
                defaultMessage="Next Task"
                id="ActiveTask.controls.nextTask.label"
                values={Object {}}
/>,
              <SvgSymbol
                sym="forward-icon"
                viewBox="0 0 20 20"
/>,
            ],
            "className": "is-text task-done-controls__next",
            "onClick": [Function],
            "title": undefined,
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "defaultMessage": "Next Task",
                "id": "ActiveTask.controls.nextTask.label",
                "values": Object {},
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "sym": "forward-icon",
                "viewBox": "0 0 20 20",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
          ],
          "type": "a",
        },
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div
            className="task-done-controls--task-status-message"
>
            undefined undefined
</div>,
          <div
            className="has-centered-children"
>
            <a
                        className="is-text task-done-controls__next"
                        onClick={[Function]}
                        title={undefined}
            >
                        <FormattedMessage
                                    defaultMessage="Next Task"
                                    id="ActiveTask.controls.nextTask.label"
                                    values={Object {}}
                        />
                        <SvgSymbol
                                    sym="forward-icon"
                                    viewBox="0 0 20 20"
                        />
            </a>
</div>,
        ],
        "className": "task-done-controls",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "undefined undefined",
            "className": "task-done-controls--task-status-message",
          },
          "ref": null,
          "rendered": "undefined undefined",
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <a
              className="is-text task-done-controls__next"
              onClick={[Function]}
              title={undefined}
>
              <FormattedMessage
                            defaultMessage="Next Task"
                            id="ActiveTask.controls.nextTask.label"
                            values={Object {}}
              />
              <SvgSymbol
                            sym="forward-icon"
                            viewBox="0 0 20 20"
              />
</a>,
            "className": "has-centered-children",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <FormattedMessage
                  defaultMessage="Next Task"
                  id="ActiveTask.controls.nextTask.label"
                  values={Object {}}
/>,
                <SvgSymbol
                  sym="forward-icon"
                  viewBox="0 0 20 20"
/>,
              ],
              "className": "is-text task-done-controls__next",
              "onClick": [Function],
              "title": undefined,
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "defaultMessage": "Next Task",
                  "id": "ActiveTask.controls.nextTask.label",
                  "values": Object {},
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "sym": "forward-icon",
                  "viewBox": "0 0 20 20",
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
            ],
            "type": "a",
          },
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
